### DISCUSSION TITLE
Clarification needed on meaning of GPUImageDataLayout

### DISCUSSION THREAD

**Discussion by trbabb** [gpuweb/gpuweb#4910] [Q&A]
I'm trying to copy a sub-rectangle of some source texture data using `queue.writeTexture()`, and it's hazy what the correct arguments for `writeTexture()` and the passed `GPUImageDataLayout` should be.

Specifically, it seems ambiguous if a given offset, origin, or byte count is referring to:

- A dimension / offset in the source data
- A dimension / offset in the *portion of the source data to be copied*
- A dimension / offset in the destination buffer

Mostly the confusion is between the first two.

For example, should `GPUImageDataLayout.rowsPerImage` contain the number of rows in the complete source buffer (starting at the pointer passed to the *data* argument), or should it count just the number of rows to be copied (presumably starting at data + offset)?

Does `GPUImageDataLayout.bytesPerRow` contain the number of bytes per row in the *original source data*, or the number of bytes per row *to be copied*? Or are these necessarily the same? If the former, what is the formula for the number of bytes per row that will actually be copied? If the latter, how is the stride specified?

I presume the `size` argument to `writeTexture()` refers to the number of texels *to be copied*, and not the extents of the original source data or the target texture?

Similar questions as above for the meaning of `offset` and its effect on where a row is considered to end.

Thanks for any clarification, and extra thanks if the docs could be updated!

**Comment by greggman** [comment]
The function is specified as `writeTexture(destination, data, dataLayout, size)`. `dataLayout` refers to `data` only.

`dataLayout.bytesPerRow` is how many bytes to advance in `data` per row. `rowsPerImage` is how many rows to advance in `data` per slice of a 3d/2d-array image. `offset` is where to start

`size` specifies how much to copy. There are no extents for the source. It's just a pile of bytes from the POV of the API.

// pseudo code
```js
writeTexture(destination, data, dataLayout, size) {
  { texture, mipLevel = 0, origin = [0,0,0] } = destination
  { offset = 0, bytesPerRow, rowsPerImage } = dataLayout (describes "data", the source)
  { width, height, depth } = size

  // how many bytes are copied per row is determined by the size and texture format
  const numBytesToCopyForRow = width * bytesPerPixelForTextureFormat(texture.format)
  
  // for each slice
  let imageOffset = offset;
  for (let z = 0; z < depth; ++z) {
  
    let rowOffset = imageOffset;
  
    // for each row
    for (let y = 0; y < height; ++y) {
  
       copy(
          from: data + rowOffset,
          to: {
            mipLevel, 
            destOrigin: [ origin.x, origin.y + y, origin.z + z ]
          }
          numBytesToCopyForRow,
       )
  
       // advance src to next row of image
       rowOffset += bytesPerRow
    }

    // advance src to next slice of image
    imageOffset += bytesPerRow * rowsPerImage;
  }
}
```  

This is all spelled out [in the spec for `writeTexture`](https://gpuweb.github.io/gpuweb/#dom-gpuqueue-writetexture) where it lists the steps for copying.

> Specifically, it seems ambiguous if a given offset, origin, or byte count is referring to:

`offset` is part of `dataLayout` so it refers to `data`
`origin` is part of `destination` so it refers to the destination
`bytesPerRow` is part of `dataLayout` so it refers to `data

>For example, should GPUImageDataLayout.rowsPerImage contain the number of rows in the complete source buffer (starting at the pointer passed to the data argument), or should it count just the number of rows to be copied (presumably starting at data + offset)?

It's just a stride to get from one slice of the image to the next. If you're only copying one slice it does not need to be specified since it's only referenced to get to the next slice.

> Does GPUImageDataLayout.bytesPerRow contain the number of bytes per row in the original source data, or the number of bytes per row to be copied? Or are these necessarily the same? If the former, what is the formula for the number of bytes per row that will actually be copied? If the latter, how is the stride specified?

`bytesPerRow` is just a stride for the source `data` to get from one row to the next. The number of bytes actually copied is determined by the texture format and width of the copy size (the last parameter to the funciton). You can look up the bytes per format [in the spec](https://gpuweb.github.io/gpuweb/#texture-format-caps). 

* note: the code above is simpilified - in particular it ignores compressed textures - the spec has the actual algorithm.
